service/controller/legacy/v29/ebs/ebs.go

```diff
@@ -128,6 +128,61 @@ func (e *EBS) DetachVolume(ctx context.Context, volumeID string, attachment Volu
 		if err != nil {
 			return microerror.Mask(err)
 		}
+	}
+
+	{
+		o := func() error {
+			i := &ec2.DescribeVolumesInput{
+				VolumeIds: []*string{
+					aws.String(volumeID),
+				},
+			}
+
+			v, err := e.client.DescribeVolumes(i)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			if len(v.Volumes) == 0 {
+				// Fall through.
+				e.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("volume %#q not found", volumeID))
+				return nil
+			}
+
+			if len(v.Volumes) > 1 {
+				// Fall through.
+				e.logger.LogCtx(ctx, "level", "debug", "message", "multiple volumes found")
+				return nil
+			}
+
+			d := v.Volumes[0]
+			if len(d.Attachments) > 1 {
+				// Shouldn't happen; log so we know if it is
+				e.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found multiple attachment for volume %#q", volumeID))
+			}
+
+			for _, a := range d.Attachments {
+				if a.State != nil {
+					attachmentStatus := *a.State
+					if attachmentStatus != "detached" {
+						return microerror.New(fmt.Sprintf("the volume %#q is still attached", volumeID))
```

Comment:
This should be a typed error which we can match against in the resource. When we then match against the returned error, we cancel the resource.